documents/repository.go

```diff
@@ -1,39 +1,195 @@
 	// Create stores the initial document
 	// If document exist, it errors out
 	Create(id []byte, model Model) error
-}
 
-// Updater interface can be implemented by any type that handles document update
-type Updater interface {
 	// Update updates the already stored document
 	// errors out when document is missing
 	Update(id []byte, model Model) error
 }
 
-// Repository should be implemented by any type that wants to store a document in key-value storage
+// Repository defines the required methods for a document repository.
+// Can be implemented by any type that stores the documents. Ex: levelDB, sql etc...
 type Repository interface {
-	Checker
-	Loader
-	Creator
-	Updater
+	// Exists checks if the id, owned by tenantID, exists in DB
+	Exists(tenantID, id []byte) bool
+
+	// Get returns the Model associated with ID, owned by tenantID
+	Get(tenantID, id []byte) (Model, error)
+
+	// Create creates the model if not present in the DB.
+	// should error out if the document exists.
+	Create(tenantID, id []byte, model Model) error
+
+	// Update strictly updates the model.
+	// Will error out when the model doesn't exist in the DB.
+	Update(tenantID, id []byte, model Model) error
+
+	// Register registers the model so that the DB can return the document without knowing the type
+	Register(model Model)
+}
+
+// levelDBRepo implements Repository using LevelDB as storage layer
+type levelDBRepo struct {
+	db     *leveldb.DB
+	models map[string]reflect.Type
+	mu     sync.RWMutex // to protect the models
+}
+
+// value is an internal representation of how levelDb stores the model.
+type value struct {
+	Type string          `json:"type"`
+	Data json.RawMessage `json:"data"`
+}
+
+// NewLevelDBRepository returns levelDb implementation of Repository
+func NewLevelDBRepository(db *leveldb.DB) Repository {
+	return &levelDBRepo{
+		db:     db,
+		models: make(map[string]reflect.Type),
+	}
+}
+
+// getKey returns tenantID+id
+func getKey(tenantID, id []byte) []byte {
+	return append(tenantID, id...)
+}
+
+// Exists returns true if the id, owned by tenantID, exists.
+func (l *levelDBRepo) Exists(tenantID, id []byte) bool {
+	key := getKey(tenantID, id)
+	res, err := l.db.Has(key, nil)
+	// TODO check this
+	if err != nil {
+		return false
+	}
+
+	return res
+}
+
+// getModel returns a new instance of the type mt.
+func (l *levelDBRepo) getModel(mt string) (Model, error) {
+	tp, ok := l.models[mt]
+	if !ok {
+		return nil, fmt.Errorf("type %s not registered", mt)
+	}
+
+	return reflect.New(tp).Interface().(Model), nil
+}
+
+// Get returns the model associated with ID, owned by tenantID.
+func (l *levelDBRepo) Get(tenantID, id []byte) (Model, error) {
+	key := getKey(tenantID, id)
+	data, err := l.db.Get(key, nil)
+	if err != nil {
+		return nil, fmt.Errorf("document missing: %v", err)
+	}
+
+	v := new(value)
+	err = json.Unmarshal(data, v)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal value: %v", err)
+	}
+
+	l.mu.RLock()
```

Comment:
Why do we need a mutex here? What would be the race condition? Just out of interest?

Reply:
This is a ReadLock which protectx access to `l.models`